File: packages/core/src/config/loader.ts
import type { IObjectStore } from "../interfaces.js"
import { SyncConfigSchema } from "./schema.js"
import type { SyncConfig } from "./schema.js"
import { validateConfig } from "./validate.js"
import { StartupError } from "../errors.js"
import { DEFAULT_CONFIG_KEY, CONTENT_TYPE_JSON } from "../constants.js"

/**
 * Load and validate a SyncConfig from storage.
 * Returns null if no config exists at the given key.
 */
export async function loadConfig(
  store: IObjectStore,
  configKey = DEFAULT_CONFIG_KEY
): Promise<SyncConfig | null> {
  const raw = await store.getString(configKey)
  if (raw === null) return null

  const parsed = SyncConfigSchema.parse(JSON.parse(raw))
  const errors = validateConfig(parsed)
  if (errors.length > 0) {
    throw new StartupError(`Invalid sync config:\n${errors.join("\n")}`)
  }
  return parsed
}

/**
 * Save a SyncConfig to storage.
 * Validates before saving; throws on invalid config.
 */
export async function saveConfig(
  store: IObjectStore,
  config: SyncConfig,
  configKey = DEFAULT_CONFIG_KEY
): Promise<void> {
  // Validate schema
  SyncConfigSchema.parse(config)

  // Semantic validation
  const errors = validateConfig(config)
  if (errors.length > 0) {
    throw new StartupError(`Invalid sync config:\n${errors.join("\n")}`)
  }

  await store.put(configKey, JSON.stringify(config, null, 2), {
    contentType: CONTENT_TYPE_JSON,
  })
}

File: servers/python/satellite_server/config/loader.py
"""Load and save sync configuration from/to object storage or JSON."""

from __future__ import annotations

import json
from pathlib import Path

from satellite_server.interfaces import IObjectStore
from satellite_server.config.schema import SyncConfig
from satellite_server.config.validate import validate_config
from satellite_server.errors import StartupError
from satellite_server.constants import DEFAULT_CONFIG_KEY, CONTENT_TYPE_JSON


def parse_config_json(raw: str) -> SyncConfig:
    """Parse and validate a SyncConfig from a JSON string.

    Accepts camelCase keys (e.g. ``storagePath``) as used in JSON config files.

    Raises :class:`StartupError` if semantic validation fails.
    """
    parsed = SyncConfig.model_validate_json(raw)
    errors = validate_config(parsed)
    if errors:
        raise StartupError(f"Invalid sync config:\n" + "\n".join(errors))
    return parsed


def load_config_file(path: str | Path) -> SyncConfig:
    """Load and validate a SyncConfig from a JSON file on disk.

    Raises :class:`FileNotFoundError` if the file does not exist, or
    :class:`StartupError` if validation fails.
    """
    return parse_config_json(Path(path).read_text(encoding="utf-8"))


async def load_config(
    store: IObjectStore,
    config_key: str = DEFAULT_CONFIG_KEY,
) -> SyncConfig | None:
    """Load and validate a SyncConfig from storage.

    Returns None if no config exists at the given key.
    """
    raw = await store.get_string(config_key)
    if raw is None:
        return None

    parsed = SyncConfig.model_validate_json(raw)
    errors = validate_config(parsed)
    if errors:
        raise StartupError(f"Invalid sync config:\n" + "\n".join(errors))
    return parsed


async def save_config(
    store: IObjectStore,
    config: SyncConfig,
    config_key: str = DEFAULT_CONFIG_KEY,
) -> None:
    """Save a SyncConfig to storage. Validates before saving."""
    errors = validate_config(config)
    if errors:
        raise StartupError(f"Invalid sync config:\n" + "\n".join(errors))

    await store.put(
        config_key,
        config.model_dump_json(by_alias=True, indent=2),
        content_type=CONTENT_TYPE_JSON,
    )
